server/block/campfire.go

```diff
@@ -0,0 +1,256 @@
+)
+
+// Campfire is a block that can be used to cook food, pacify bees, act as a spread-proof light source, smoke signal or
+// damaging trap block.
+type Campfire struct {
+	bass
+	sourceWaterDisplacer
+
+	// Items represents the items in the campfire that are being cooked.
+	Items [4]CampfireItem
+	// Facing represents the direction that the campfire is facing.
+	Facing cube.Direction
+	// Extinguished is true if the campfire was extinguished by a water source.
+	Extinguished bool
+	// Type represents the type of Campfire, currently there are Normal and Soul campfires.
+	Type FireType
+}
+
+// CampfireItem holds data about the items in the campfire.
+type CampfireItem struct {
+	// Item is a specific item being cooked on top of the campfire.
+	Item item.Stack
+	// Time is the countdown of ticks until the food item is cooked (when 0).
+	Time time.Duration
+}
+
+// Model ...
+func (c Campfire) Model() world.BlockModel {
+	return model.Campfire{}
+}
+
+// SideClosed ...
+func (c Campfire) SideClosed(cube.Pos, cube.Pos, *world.World) bool {
+	return false
+}
+
+// BreakInfo ...
+func (c Campfire) BreakInfo() BreakInfo {
+	return newBreakInfo(5, alwaysHarvestable, axeEffective, func(t item.Tool, enchantments []item.Enchantment) []item.Stack {
+		var drops []item.Stack
+		if hasSilkTouch(enchantments) {
+			drops = append(drops, item.NewStack(c, 1))
+		} else {
+			switch c.Type {
+			case NormalFire():
+				drops = append(drops, item.NewStack(item.Charcoal{}, 2))
+			case SoulFire():
+				drops = append(drops, item.NewStack(SoulSoil{}, 1))
+			}
+		}
+		for _, v := range c.Items {
+			if !v.Item.Empty() {
+				drops = append(drops, v.Item)
+			}
+		}
+		return drops
+	})
+}
+
+// LightEmissionLevel ...
+func (c Campfire) LightEmissionLevel() uint8 {
+	if c.Extinguished {
+		return 0
+	}
+	return c.Type.LightLevel()
+}
+
+// Ignite ...
+func (c Campfire) Ignite(pos cube.Pos, w *world.World) bool {
+	w.PlaySound(pos.Vec3(), sound.Ignite{})
+	if _, ok := w.Liquid(pos); ok {
+		return true
+	}
+
+	c.Extinguished = false
+	w.SetBlock(pos, c, nil)
+	return true
+}
+
+// Splash ...
+func (c Campfire) Splash(w *world.World, pos cube.Pos, p potion.Potion) {
+	if p != potion.Water() {
+		// Water is the only potion that can extinguish a campfire.
+		return
+	}
+
+	w.PlaySound(pos.Vec3Centre(), sound.FireExtinguish{})
+	c.Extinguished = true
+	w.SetBlock(pos, c, nil)
+}
+
+// Activate ...
+func (c Campfire) Activate(pos cube.Pos, _ cube.Face, w *world.World, u item.User, ctx *item.UseContext) bool {
+	if held, _ := u.HeldItems(); !held.Empty() {
+		if rawFood, ok := held.Item().(item.Smeltable); ok && rawFood.SmeltInfo().Food {
+			for i, it := range c.Items {
+				if it.Item.Empty() {
+					c.Items[i] = CampfireItem{
+						Item: held.Grow(-held.Count() + 1),
+						Time: time.Second * 30,
+					}
+
+					ctx.SubtractFromCount(1)
+
+					w.PlaySound(pos.Vec3Centre(), sound.ItemAdd{})
+					w.SetBlock(pos, c, nil)
+					return true
+				}
+			}
+		}
+	}
```

Comment:
This is some super deep nesting. Instead, consider early returns like:
```go
held, _ := u.HeldItems()
if held.Empty() {
  return false
}
if rawFood, ok := held.Item().(item.Smeltable); !ok || !rawFood.SmeltInfo().Food {
  return false
}
...
```
This makes it much easier to read.